docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,363 @@
+# CFEngine custom promise types
+
+Custom promise types can be added as _Promise modules_.
+These enable users to implement their own promise types for managing new resources.
+They can be added without a binary upgrade, and be shared between users.
+
+## Using custom promise types
+
+A new top level `promise` block (similar to `body`) is used to add new promise types.
+After adding a promise type, it is used in the same way as built-in promise types, there is no special syntax.
+
+### Example of using a promise module
+
+```cf3
+promise agent git
+{
+  interpreter => "/usr/bin/python3";
+  path => "/var/cfengine/inputs/modules/promises/git.py";
+}
+
+bundle agent main
+{
+  git:
+    "/opt/cfengine/masterfiles"
+      repo => "https://github.com/cfengine/masterfiles";
+}
+```
+
+## Evaluation details
+
+Common attributes and class guards are handled by the agent, and not sent to the promise module.
```

Comment:
It would nice to have a `jq` that would produce a list of the common attributes from `cf-promsies --syntax-description` for easy reference.

Reply:
Closed as off topic.